File: ADA/Sort/06-Merge-Sort/MergeSort.c
#include <stdio.h>
#include <stdlib.h>

void merge(int arr[], int start, int mid, int end)
{

    int mergerd[end - start + 1];

    int leftPointer = start;
    int rightPointer = mid + 1;
    int x = 0;

    while (leftPointer <= mid && rightPointer <= end)
    {
        if (arr[leftPointer] <= arr[rightPointer])
        {
            mergerd[x] = arr[leftPointer];
            x++;
            leftPointer++;
        }
        else
        {
            mergerd[x] = arr[rightPointer];
            x++;
            rightPointer++;
        }
    }

    while (leftPointer <= mid)
    {
        mergerd[x] = arr[leftPointer];
        x++;
        leftPointer++;
    }
    while (rightPointer <= end)
    {
        mergerd[x] = arr[rightPointer];
        x++;
        rightPointer++;
    }
    int i = start, // index for arr
        j = 0;     // index for mergerd

    while (j < end - start + 1)
    {
        arr[i++] = mergerd[j++];
    }
}

void mergeSort(int arr[], int start, int end)
{
    if (start >= end)
        return;

    int mid = start + (end - start) / 2;

    mergeSort(arr, start, mid);
    mergeSort(arr, mid + 1, end);

    merge(arr, start, mid, end);
}

void printArray(int arr[], int size)
{
    int i;
    for (i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

void main()
{
    int arr[] = {2, 5, 4, 1, 3};
    int n = sizeof(arr) / sizeof(arr[0]);
    printf("Original array	: ");
    printArray(arr, n);

    mergeSort(arr, 0, n - 1);

    printf("\n\nSorted array	: ");
    printArray(arr, n);
}

File: ADA/Sort/06-Merge-Sort/MergeSort.java
class MergeSort {

    static void merge(int arr[], int start, int mid, int end) {

        int merged[] = new int[end - start + 1];
        
        int leftPointer = start;
        int rightPointer = mid + 1;
        int x = 0;
        while (leftPointer <= mid && rightPointer <= end) {
            if (arr[leftPointer] <= arr[rightPointer]) {
                merged[x++] = arr[leftPointer++];
            } else {
                merged[x++] = arr[rightPointer++];
            }
        }
        while (leftPointer <= mid) {
            merged[x++] = arr[leftPointer++];
        }
        while (rightPointer <= end) {
            merged[x++] = arr[rightPointer++];
        }
        for (int i = start, j = 0; j < end - start + 1; i++, j++) {
            arr[i] = merged[j];
        }

    }

    static void mergeSort(int arr[], int si, int ei) {
        if (si >= ei)
            return;
        int mid = si + (ei - si) / 2;

        mergeSort(arr, si, mid);
        mergeSort(arr, mid + 1, ei);

        merge(arr, si, mid, ei);
    }

    static void printArray(int arr[]) {
        for (int i = 0; i < arr.length; i++)
            System.out.print(arr[i] + " ");
    }

    public static void main(String args[]) {
        int arr[] = { 5, 4, 3, 2, 1 };
        System.out.print("Original array	: ");
        printArray(arr);

        mergeSort(arr, 0, arr.length - 1);

        System.out.print("\n\nSorted array	: ");
        printArray(arr);
    }
}